### PR TITLE
FEAT: async firebase

### PR DESCRIPTION
**Pull request by avo2003**

### CODE REVIEW

File: src/middlewares/checkJwt.ts [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
        const req = context.switchToHttp().getRequest()
        const res = context.switchToHttp().getResponse()
-       const token = req.headers.authorization.split(' ');
+       const token = req.headers.authorization.split(' ')[1];
+       
          try {
-            const jwtPayload = jwt.verify(token[1], process.env.JWT_ACCESS) as JwtPayload;
-            res.locals.token  = jwtPayload.id;
+              const decodedIdToken = await admin.auth().verifyIdToken(token)
+                res.locals.userUid = decodedIdToken.uid
```
Comment (harutyunyan611) [inline src/middlewares/checkJwt.ts]: Spaces

File: src/users/controller/userController.ts [hunk 2]
```diff
@@ -20,27 +18,18 @@ import { UsersService } from "../service/userService";
...
   @Inject()
   usersService: UsersService;
   @UseInterceptors(ClassSerializerInterceptor)
   @Post('create')
-  async create(@Body() createUserDto: CreateUserDto,): Promise<User> {
-    const hashedPassword = await bcrypt.hash(createUserDto.password, +process.env.BCRYPT_KEY);
-    createUserDto.password = hashedPassword;
-    createUserDto.avatar=process.env.AVATAR_URL;
+  async create(@Body() createUserDto: CreateUserDto): Promise<User> {
+     createUserDto.avatar=process.env.AVATAR_URL;
```
Comment (harutyunyan611) [inline src/users/controller/userController.ts]: Why in the controller?

File: src/users/controller/userController.ts [hunk 3]
```diff
@@ -52,31 +41,11 @@ export class UsersController {
...
-      {secret: process.env.JWT_ACCESS, expiresIn: process.env.LOGIN_REFRESH_TOKEN_TIME});
-    return  {accesToken,refreshToken}
-  }
-
   @UseInterceptors(ClassSerializerInterceptor)
   @UseGuards(AuthGuard)
   @ApiBearerAuth()
   @Get('me')
-   findOne(@Token() data: number): Promise<User> {
+   findOne(@Token() data: string): Promise<User> {
```
Comment (harutyunyan611) [inline src/users/controller/userController.ts]: What is data?

File: src/users/controller/userController.ts [hunk 4]
```diff
@@ -90,48 +59,17 @@ export class UsersController {
...
-      const jwtPayload: jwt.JwtPayload | string = jwt.verify(reset.token, process.env.JWT_ACCESS_FORGOT) as jwt.JwtPayload;
-      if(reset.password === reset.confirmPass){
-        const hashedPassword = await bcrypt.hash(reset.password, +process.env.BCRYPT_KEY);
-          this.usersService.resetFindeOne(jwtPayload.id,hashedPassword);
-          return "pass is changed";
-      }}
-    catch(error){
-      return error ;
-    }
+  remove(@Param('id') id: number,@Token() data: string): Promise<User> {
```
Comment (harutyunyan611) [inline src/users/controller/userController.ts]: Same here

File: src/users/controller/userController.ts [hunk 5]
```diff
@@ -90,48 +59,17 @@ export class UsersController {
...
+    return this.usersService.remove(id,data);
   }
 
   @Patch('avatar')
   @UseGuards(AuthGuard)
   @ApiBearerAuth()
   @UseInterceptors(FileInterceptor('avatar'))
-   changeUserImg(@UploadedFile() file: Express.Multer.File,@Token() data: number,
+   changeUserImg(@UploadedFile() file: Express.Multer.File,@Token() data: string,
```
Comment (harutyunyan611) [inline src/users/controller/userController.ts]: Here

File: src/users/controller/userController.ts [hunk 6]
```diff
@@ -140,8 +78,8 @@ export class UsersController {
   @UseGuards(AuthGuard)
   @ApiBearerAuth()
   @Patch(':id/salary')
-  changeSalary(
-    @Token() data: number,
+   changeSalary(
```
Comment (harutyunyan611) [inline src/users/controller/userController.ts]:
```suggestion
  changeSalary(
```

File: src/users/controller/userController.ts [hunk 7]
```diff
@@ -140,8 +78,8 @@ export class UsersController {
...
+    @Token() data: string,
```
Comment (harutyunyan611) [inline src/users/controller/userController.ts]: WHy data?

File: src/users/dto/userCreateDto.ts [hunk 8]
```diff
@@ -11,21 +11,24 @@ export class CreateUserDto extends User {
   @IsString()
   readonly firstName: string;
 
-  @ApiProperty()
-  @IsString()
-  readonly lastName: string;
+  @IsOptional()
+  authUid: string;
```
Comment (harutyunyan611) [inline src/users/dto/userCreateDto.ts:15]: Why optional?

File: src/users/repository/userRepository.ts [hunk 9]
```diff
@@ -32,8 +32,11 @@ export class UserRepository {
...
 
-  async remove(id: number): Promise<User> {
-    const user = await this.findOne(id);
-    return this.userRepository.remove(user);
+  async remove(id: number,data:string): Promise<User> {
+    const user = await this.findOne(data);
+    const removeUserId = await this.userRepository.findOne(id)
+    if(user.isAdmin){
+      await dbAuth.deleteUser(removeUserId.authUid);
+    return this.userRepository.remove(removeUserId);}
```
Comment (harutyunyan611) [inline src/users/repository/userRepository.ts]: Spacing meseed up

File: src/users/service/userService.ts [hunk 10]
```diff
@@ -53,27 +47,15 @@ export class UsersService {
     return user;
   }
 
-  remove(id: number): Promise<User> {
-    const user =this.usersRepository.remove(id);
+  remove(id: number,data:string): Promise<User> {
+    const user =this.usersRepository.remove(id,data);
     if (!user) {
```
Comment (harutyunyan611) [inline src/users/service/userService.ts]: What is data?

**Review by harutyunyan611 — CHANGES_REQUESTED** [review]

File: src/users/controller/userController.ts [hunk 11]
```diff
@@ -90,61 +58,30 @@ export class UsersController {
...
-      if(reset.password === reset.confirmPass){
-        const hashedPassword = await bcrypt.hash(reset.password, +process.env.BCRYPT_KEY);
-          this.usersService.resetFindeOne(jwtPayload.id,hashedPassword);
-          return "pass is changed";
-      }}
-    catch(error){
-      return error ;
-    }
+  remove(@Param('id') id: number,@Token() uid: string): Promise<User> {
+    return this.usersService.remove(id,uid);
```
Comment (ArthurO2002) [inline src/users/controller/userController.ts]: fix tabulation on each file

File: src/users/dto/userCreateDto.ts [hunk 12]
```diff
@@ -11,21 +11,24 @@ export class CreateUserDto extends User {
...
   @ApiProperty()
-  @IsEmail()
-  readonly email: string;
+  @IsString()
+  password: string;
 
   @ApiProperty()
   @IsString()
-   password: string;
+  readonly lastName: string;
```
Comment (ArthurO2002) [inline src/users/dto/userCreateDto.ts]: ?????!!!!

File: src/users/dto/userCreateDto.ts [hunk 13]
```diff
@@ -11,21 +11,24 @@ export class CreateUserDto extends User {
...
   avatar: string;
 
+  @ApiProperty()
+  @IsEmail()
+  readonly email: string;
+
   @IsEmpty()
   readonly events: Event[];
-}
+}
```
Comment (ArthurO2002) [inline src/users/dto/userCreateDto.ts]: Why are events readOnly? @Zt0

File: src/users/repository/userRepository.ts [hunk 14]
```diff
@@ -32,8 +32,11 @@ export class UserRepository {
...
-  async remove(id: number): Promise<User> {
-    const user = await this.findOne(id);
-    return this.userRepository.remove(user);
+  async remove(id: number,data:string): Promise<User> {
+    const user = await this.findOne(data);
+    const removeUserId = await this.userRepository.findOne(id)
+    if(user.isAdmin){
+      await dbAuth.deleteUser(removeUserId.authUid);
+      return this.userRepository.remove(removeUserId);}
   }
```
Comment (ArthurO2002) [inline src/users/repository/userRepository.ts]: tabulation

Comment (ArthurO2002) [reply src/users/repository/userRepository.ts]: If user will not be admin, they will receive nothing?

File: src/users/service/userService.ts [hunk 15]
```diff
@@ -53,27 +48,15 @@ export class UsersService {
     return user;
   }
 
-  remove(id: number): Promise<User> {
-    const user =this.usersRepository.remove(id);
+  remove(id: number,uid:string): Promise<User> {
+    const user =this.usersRepository.remove(id,uid);
     if (!user) {
       throw new NotFoundException(`User with ID=${id} not found`);
     }
```
Comment (ArthurO2002) [inline src/users/service/userService.ts]: Test and attach screenshot, how is it working))))


add await to make this work

File: src/users/controller/userController.ts [hunk 16]
```diff
@@ -52,32 +40,12 @@ export class UsersController {
...
-  }
-
   @UseInterceptors(ClassSerializerInterceptor)
   @UseGuards(AuthGuard)
   @ApiBearerAuth()
   @Get('me')
-   findOne(@Token() data: number): Promise<User> {
-    return  this.usersService.findOne(data)
+   findOne(@Token() uid: string): Promise<User> {
+    return  this.usersService.findOne(uid)
```
Comment (ArthurO2002) [inline src/users/controller/userController.ts]: How this will work, if it does not have await

File: src/users/controller/userController.ts (hunk 11 above)
Comment (ArthurO2002) [reply src/users/controller/userController.ts]: Without await again?

File: src/users/controller/userController.ts [hunk 17]
```diff
@@ -90,61 +58,30 @@ export class UsersController {
...
   @ApiBearerAuth()
   @Patch(':id/salary')
   changeSalary(
-    @Token() data: number,
+    @Token() uid: string,
     @Body() body: {salary: number},
     @Param('id') id:number
   ): Promise<User> {
-    return  this.usersService.changeSalary(data,body.salary,id)
+    return  this.usersService.changeSalary(uid,body.salary,id)
```
Comment (ArthurO2002) [inline src/users/controller/userController.ts]: fix tabulation

File: src/users/repository/userRepository.ts [hunk 18]
```diff
@@ -19,8 +19,8 @@ export class UserRepository {
     return this.userRepository.find({relations: ["events"],});
   }
 
-   findOne(id: number): Promise<User> {
-    const user =  this.userRepository.findOne(id, {relations: ["events"]});
+  async findOne(id: string): Promise<User> {
+    const user = await this.userRepository.findOne( {relations: ["events"], where:{authUid: id }});
     return user;
```
Comment (ArthurO2002) [inline src/users/repository/userRepository.ts]: unnecessary const

File: src/users/service/userService.ts [hunk 19]
```diff
@@ -17,34 +18,28 @@ export class UsersService {
...
   usersRepository: UserRepository;
   async create(createUserDto: CreateUserDto): Promise<User> {
-
+    const auth =  await dbAuth.createUser({ email:createUserDto.email, password:createUserDto.password })
+    createUserDto.authUid=auth.uid
+    createUserDto.avatar=process.env.AVATAR_URL;
     const user = await this.usersRepository.create(createUserDto);
     if (!user) {
-      throw new NotFoundException(`how have wrong schema`);
+      throw new NotFoundException(`you have wrong schema`);
```
Comment (ArthurO2002) [inline src/users/service/userService.ts]: what schema, error is not understandable

Comment (ArthurO2002) [reply src/users/service/userService.ts]: throw ner error with statusCode 405 and with correct error message

File: src/users/service/userService.ts [hunk 20]
```diff
@@ -17,34 +18,28 @@ export class UsersService {
...
   }
 
   findAll(): Promise<Array<User>> {
     return this.usersRepository.findAll();
   }
 
-  async findOne(id: number): Promise<User> {
-    const user = await this.usersRepository.findOne(id);
+  async findOne(authUid: string): Promise<User> {
+    const user =  this.usersRepository.findOne(authUid);
```
Comment (ArthurO2002) [inline src/users/service/userService.ts]: use await

File: src/users/service/userService.ts [hunk 21]
```diff
@@ -53,27 +48,15 @@ export class UsersService {
     return user;
   }
 
-  remove(id: number): Promise<User> {
-    const user =this.usersRepository.remove(id);
+  remove(id: number,uid:string): Promise<User> {
+    const user =this.usersRepository.remove(id,uid);
```
Comment (ArthurO2002) [inline src/users/service/userService.ts]: fix tabulation

File: src/users/service/userService.ts [hunk 22]
```diff
@@ -53,27 +48,15 @@ export class UsersService {
...
-      throw new NotFoundException(`User with ID=${id} not found`);
-    }
-    return this.userRepository.save(user);
-
-  }
-
-  async changeSalary(userId: number,salary:number,id:number): Promise<User> {
+  async changeSalary(userId: string,salary:number,id:number): Promise<User> {
     const user = await this.usersRepository.findOne(userId);
     if(user.isAdmin){
```
Comment (ArthurO2002) [inline src/users/service/userService.ts]: configure eslint and prettier for your webstorm

File: src/users/service/userService.ts [hunk 23]
```diff
@@ -87,14 +70,14 @@ export class UsersService {
     }
   }
 
-  async uploadImageToCloudinary(file: Express.Multer.File,id: number) {
+  async uploadImageToCloudinary(file: Express.Multer.File,id: string) {
     const userId = await this.usersRepository.findOne(id);
     if(userId.avatarPublicId){
       await this.cloudinary.deleteImg(userId.avatarPublicId);
```
Comment (ArthurO2002) [inline src/users/service/userService.ts]: tabulation

**Review by ArthurO2002 — CHANGES_REQUESTED** [review]
@harutyunyan611 to confirm

File: src/users/dto/userCreateDto.ts (hunk 12 above)
Comment (ArthurO2002) [reply src/users/dto/userCreateDto.ts]: why is lastname readonly

File: src/users/controller/userController.ts (hunk 11 above)
Comment (avo2003) [reply src/users/controller/userController.ts:64]: done

File: src/users/controller/userController.ts [hunk 24]
```diff
@@ -52,32 +40,12 @@ export class UsersController {
...
-  }
-
   @UseInterceptors(ClassSerializerInterceptor)
   @UseGuards(AuthGuard)
   @ApiBearerAuth()
   @Get('me')
-   findOne(@Token() data: number): Promise<User> {
-    return  this.usersService.findOne(data)
+  async findOne(@Token() uid: string): Promise<User> {
+    return await this.usersService.findOne(uid);
```
Comment (harutyunyan611) [inline src/users/controller/userController.ts]: Why async await

File: src/users/controller/userController.ts [hunk 25]
```diff
@@ -90,61 +58,32 @@ export class UsersController {
...
-      if(reset.password === reset.confirmPass){
-        const hashedPassword = await bcrypt.hash(reset.password, +process.env.BCRYPT_KEY);
-          this.usersService.resetFindeOne(jwtPayload.id,hashedPassword);
-          return "pass is changed";
-      }}
-    catch(error){
-      return error ;
-    }
+  async remove(@Param('id') id: number, @Token() uid: string): Promise<User> {
+    return await this.usersService.remove(id, uid);
```
Comment (harutyunyan611) [inline src/users/controller/userController.ts]: Why async await?

File: src/users/dto/userCreateDto.ts (hunk 8 above)
Comment (harutyunyan611) [reply src/users/dto/userCreateDto.ts:15]: Then why in DTO? Can we remove it?

File: src/users/dto/userGetDto.ts [hunk 26]
```diff
@@ -1,10 +1,10 @@
-import { User } from "../entity/user";
+import { User } from '../entity/user';
 
 export const UserGetDto = (user: User): User => ({
   id: user.id,
   firstName: user.firstName,
   lastName: user.lastName,
-  password: user.password,
+  authUid: user.authUid,
```
Comment (harutyunyan611) [inline src/users/dto/userGetDto.ts]: I think same here, maybe not

File: src/users/repository/userRepository.ts [hunk 27]
```diff
@@ -1,27 +1,30 @@
...
 
 @Injectable()
 export class UserRepository {
   @InjectRepository(User)
   userRepository: Repository<User>;
 
-   create(createUserDto: CreateUserDto): Promise<User> {
-    const user =  this.userRepository.create(createUserDto);
+  create(createUserDto: CreateUserDto): Promise<User> {
+    const user = this.userRepository.create(createUserDto);
```
Comment (harutyunyan611) [inline src/users/repository/userRepository.ts]: Can you check the type of this? If it's promise we should add await

File: src/users/repository/userRepository.ts [hunk 28]
```diff
@@ -1,27 +1,30 @@
...
-    return this.userRepository.find({relations: ["events"],});
+    return this.userRepository.find({ relations: ['events'] });
   }
 
-   findOne(id: number): Promise<User> {
-    const user =  this.userRepository.findOne(id, {relations: ["events"]});
-    return user;
+  async findOne(id: string): Promise<User> {
+    return await this.userRepository.findOne({
+      relations: ['events'],
```
Comment (harutyunyan611) [inline src/users/repository/userRepository.ts]: Why async await?

File: src/users/repository/userRepository.ts [hunk 29]
```diff
@@ -32,8 +35,17 @@ export class UserRepository {
...
 
-  async remove(id: number): Promise<User> {
-    const user = await this.findOne(id);
-    return this.userRepository.remove(user);
+  async remove(id: number, uid: string): Promise<User> {
+    const user = await this.findOne(uid);
+    const removeUserId = await this.userRepository.findOne(id);
+    if (user.isAdmin) {
+      await dbAuth.deleteUser(removeUserId.authUid);
+      return await this.userRepository.remove(removeUserId);
```
Comment (harutyunyan611) [inline src/users/repository/userRepository.ts]: Why return await?

File: src/users/repository/userRepository.ts [hunk 30]
```diff
@@ -32,8 +35,17 @@ export class UserRepository {
...
+    }
+    if (!user) {
```
Comment (harutyunyan611) [inline src/users/repository/userRepository.ts]: Doesn't make sense, we're checking if a user is an admin, and then checking if the user exists. I think logic should be vice versa

File: src/users/service/userService.ts (hunk 19 above)
Comment (harutyunyan611) [reply src/users/service/userService.ts]: What is schema?

Comment (harutyunyan611) [reply src/users/service/userService.ts]: Why not found the error on creating the user?

File: src/users/service/userService.ts [hunk 31]
```diff
@@ -17,86 +23,77 @@ export class UsersService {
...
-   findOneEmail(condition: string): Promise<User> {
-    return this.userRepository.findOne({
-      where: [
-        { email: condition}
-      ],});
-  }
-
   update(id: number, updateUserDto: UpdateUserDto): Promise<User> {
-    const user=this.usersRepository.update(id, updateUserDto);
+    const user = this.usersRepository.update(id, updateUserDto);
```
Comment (harutyunyan611) [inline src/users/service/userService.ts]: Where is async await?

File: src/users/service/userService.ts [hunk 32]
```diff
@@ -17,86 +23,77 @@ export class UsersService {
...
-      password:pass
-    });
+  async remove(id: number, uid: string): Promise<User> {
+    const user = await this.usersRepository.remove(id, uid);
     if (!user) {
       throw new NotFoundException(`User with ID=${id} not found`);
     }
-    return this.userRepository.save(user);
-
+    return user;
```
Comment (harutyunyan611) [inline src/users/service/userService.ts]: Can we ensure what's returning remove function? Maybe we can wrap it with try catch

File: src/utils/sendEmail.ts [hunk 33]
```diff
@@ -1,22 +1,26 @@
...
-        service:'gmail',
-        auth: {
-            user: process.env.email,
-            pass: process.env.emailPass,
-        },
-    });
+  const transporter = nodemailer.createTransport({
+    service: 'gmail',
+    auth: {
+      user: process.env.email,
```
Comment (harutyunyan611) [inline src/utils/sendEmail.ts:8]: Email camelcase

File: src/utils/sendEmail.ts [hunk 34]
```diff
@@ -1,22 +1,26 @@
 import * as nodemailer from 'nodemailer';
 // const ejs = require ('ejs')
```
Comment (harutyunyan611) [inline src/utils/sendEmail.ts]: Remove comment

**Review by harutyunyan611 — CHANGES_REQUESTED** [review]

**Review by harutyunyan611 — APPROVED** [review]

**Pull request merged**